### PR TITLE
openapi(dm): support get/update cluster topology info

### LINKED ISSUE #5095: support get/update cluster topo info

### PR DESCRIPTION
**Pull request by Ehco1996**
### What problem does this PR solve?

Issue Number: close #5095

### What is changed and how it works?
* as title
* migrate openapi view test to testify
* set real master addr in swagger so user now can play the with the openapi (use **Try it out** butten)in docs view
截屏2022-04-02 11 17 28

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update key monitor metrics in both TiCDC document and official document

### Release note 

```release-note
`support get/update cluster topology info with openapi`.
```

### CODE REVIEW

**Comment by Ehco1996** [comment]
could you please take another look for this pr ? thanks 🧡🧡🧡
/cc @WizardXiao @buchuitoudegou

File: dm/dm/master/openapi_view.go [hunk 1]
```diff
@@ -85,10 +85,26 @@ func (s *Server) InitOpenAPIHandles() error {
...
 func (s *Server) GetDocJSON(c *gin.Context) {
+	var masterURL string
+	if info, err := s.getClusterInfo(c.Request.Context()); err != nil {
+		_ = c.Error(err)
+		return
+	} else if info.Topology.MasterTopologyList != nil && len(*info.Topology.MasterTopologyList) > 0 {
+		masterTopos := *info.Topology.MasterTopologyList
+		schema := "http"
+		if useTLS.Load() {
+			schema = "https"
```
Comment (buchuitoudegou) [inline dm/dm/master/openapi_view.go:96]:
```suggestion
			protocol = "https"
```

File: dm/dm/master/openapi_controller.go [hunk 2]
```diff
@@ -33,6 +35,39 @@ import (
...
+func (s *Server) getClusterInfo(ctx context.Context) (*openapi.GetClusterInfoResponse, error) {
+	info := &openapi.GetClusterInfoResponse{}
+	info.ClusterId = s.ClusterID()
+
+	resp, err := s.etcdClient.Get(ctx, dmcommon.ClusterTopologyKey)
+	if err != nil {
+		return nil, err
+	}
+
+	// already setted by tiup, load to info
```
Comment (buchuitoudegou) [inline dm/dm/master/openapi_controller.go:48]:
```suggestion
	// already set by tiup, load to info
```

**Review by buchuitoudegou — COMMENTED** [review]
lgtm (lame at golang server-end programming😭)

File: dm/openapi/swaggerui.go [hunk 3]
```diff
@@ -111,16 +111,3 @@ func GetSwaggerHTML(config *SwaggerConfig) (html string, err error) {
 	}
 	return buf.String(), nil
 }
-
-// GetSwaggerJSON returns the swagger json.
-func GetSwaggerJSON() ([]byte, error) {
```
Comment (WizardXiao) [inline dm/openapi/swaggerui.go:116]: `GetSwaggerJSON` seems to have no impact，but why was it deleted.

File: dm/dm/master/openapi_view.go [hunk 4]
```diff
@@ -165,6 +181,31 @@ func (s *Server) DMAPIOfflineWorkerNode(c *gin.Context, workerName string) {
...
+	info, err := s.getClusterInfo(c.Request.Context())
+	if err != nil {
+		_ = c.Error(err)
+		return
+	}
+	c.IndentedJSON(http.StatusOK, info)
+}
+
+// DMAPIGetClusterInfo return cluster id of dm cluster url is: (PUT /api/v1/cluster/info).
+func (s *Server) DMAPIUpdateClusterInfo(c *gin.Context) {
```
Comment (WizardXiao) [inline dm/dm/master/openapi_view.go:195]: Just a question，it seems that only `updateClusterInfo` will put info to etcd,  so the topology will be accurate when use openapi, and it should be called by front-end?

**Review by WizardXiao — COMMENTED** [review]
rest lgtm, just have some question.

File: dm/openapi/swaggerui.go (hunk 3 above)
Comment (Ehco1996) [reply dm/openapi/swaggerui.go:116]: GetSwaggerJSON is uesd with `echo` when openapi was introduced, as we use `gin` now, this func is no needed

File: dm/dm/master/openapi_view.go (hunk 4 above)
Comment (Ehco1996) [reply dm/dm/master/openapi_view.go:195]: this info is mantained by tiup, and tiup can't write our etcd directly, so we add a openapi for it, of course users can write this info, but they should ensure the accuracy of this info themselves

**Review by WizardXiao — APPROVED** [review]

**Comment by Ehco1996** [comment]
ptal @buchuitoudegou

**Review by buchuitoudegou — APPROVED** [review]
lgtm

**Comment by lance6716** [comment]
you can add the label require-LGT1. BTW @buchuitoudegou should get the review priviledge https://github.com/pingcap/community/pull/653 3 days later

**Pull request merged**